### PR TITLE
iTunes: Unit-test XML importer on actual library XMLs

### PR DESCRIPTION
**Pull request by fwcd**
This branch adds a unit test verifying that the XML importer performs correctly on real music library XMLs generated with

- Apple Music 1.3.5.3 on macOS 13.4
- iTunes 12.12.8.2 on Windows 11

To do:

- [x] Test imported tracks
- [x] Test imported playlist relations
- [x] Test imported playlists
- [x] Test generated `TreeItem`

### CODE REVIEW

File: src/test/itunesxmlimportertest.cpp [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+                    .year = 1979,
+                    .duration = 208,
+                    .location = musicRoot +
+                            "/AC_DC/Highway to Hell/01 Highway to Hell.m4a",
+                    .rating = 0,
+                    .comment = "",
+                    .trackNumber = 1,
+                    .bpm = 0,
+                    .bitrate = 256,
+            }));
```
Comment (fwcd) [inline src/test/itunesxmlimportertest.cpp:81]: We have a bit of code duplication here with the iTunes XML test below. I have used the same tracks and playlists in the hope of being able to share some logic with the other test. Unfortunately, however, the indexing is different, also there are some minor differences in automatically generated playlists, therefore I believe we cannot (easily) avoid this. Suggestions welcome.

**Comment by fwcd** [comment]
@Swiftb0y Sorry for pinging again, but this will probably conclude the series of iTunes XML PRs (for now) and having the test would be great to make sure we don't accidentally break anything in the future.

File: src/library/itunes/itunesdao.h [hunk 2]
```diff
@@ -59,6 +68,11 @@ class ITunesDAO : public DAO {
     QHash<int, QString> m_playlistNameById;
     std::multimap<int, int> m_playlistIdsByParentId;
 
+    // Keeps track of whether the database has been initialized.
+    // In tests the database is not used, so the importer will
+    // only be used to construct a TreeItem.
+    bool m_isDatabaseInitialized = false;
```
Comment (Swiftb0y) [inline src/library/itunes/itunesdao.h:74]: can you elaborate? How is this related to the tests?

File: src/library/itunes/itunesdao.cpp [hunk 3]
```diff
@@ -31,46 +58,53 @@ void ITunesDAO::initialize(const QSqlDatabase& database) {
...
-            track.trackNumber > 0 ? QVariant(track.trackNumber) : QVariant());
-    query.bindValue(":bpm", track.bpm);
-    query.bindValue(":bitrate", track.bitrate);
-
-    if (!query.exec()) {
-        LOG_FAILED_QUERY(query);
-        return false;
+    if (m_isDatabaseInitialized) {
+        QSqlQuery& query = m_insertTrackQuery;
+
```
[the remark this answers is not in the document]
Comment (fwcd) [reply src/library/itunes/itunesdao.cpp:68]: In production usage yes, but in tests we explicitly use this. Perhaps there's some cleaner way to express that without too much duplication?

File: src/library/itunes/itunesdao.cpp [hunk 4]
```diff
@@ -84,30 +118,34 @@ bool ITunesDAO::importPlaylistRelation(int parentId, int childId) {
...
+            return false;
+        }
     }
 
     return true;
 }
 
 bool ITunesDAO::applyPathMapping(const ITunesPathMapping& pathMapping) {
-    QSqlQuery& query = m_insertPlaylistTrackQuery;
+    if (m_isDatabaseInitialized) {
```
[the remark this answers is not in the document]
Comment (fwcd) [reply src/library/itunes/itunesdao.cpp:138]: See the comment above regarding the testability, otherwise I'd agree (and this is probably mostly an artifact from `DAO` being designed the way it is)

File: src/library/itunes/itunesdao.h (hunk 2 above)
Comment (fwcd) [reply src/library/itunes/itunesdao.h:74]: Aside from making db calls, the `importPlaylist` method also updates the name-id mapping and the `importPlaylistRelation` updates the parent-child mapping, both of which are needed to construct the `TreeItem` later on. Since we don't need/want the db calls during testing, I'm using this flag currently, which (kind of unintentionally) doubles as a safety check.

There are other approaches I've thought of, such as splitting up db logic and bookkeeping into separate methods, but those make the abstraction leakier and place more of a burden on the caller of the DAO, so I'm not sure whether that would be a better option.

**Comment by fwcd** [comment]
Any updates on this?

**Comment by Swiftb0y** [comment]
Sorry, I appreciate your hard work but I don't have the time right now to review this further. Can someone else please take over?

**Comment by daschuer** [comment]
iTunes is still working in my Windows V-Box.

**Review by daschuer — APPROVED** [review]
LGTM, Thank you.
